### ISSUE TITLE
[Feature Request] Filter the documents

### ISSUE THREAD

**Issue by sebojolais** [OpenBoard-org/OpenBoard#1141]
**Is your feature request related to a problem? Please describe.**
My use-case of OB leads me to got lot of documents that are not always sorted in folders. So I have a very long list of document in the `document mode`. Some of them are rename with a name that make sense for me.
Currently I am not able to find quickly a document with an meaning name due to the list size. 

**Describe the solution you'd like**
If we add a `line edit` to **filter** the documents list, it will be very convenient.
We can image something like this:

https://github.com/user-attachments/assets/1de22a18-73e9-4b2a-aa66-f47e292cf94f

By this way, the list is shorter and I am able to find the right document.

**Describe alternatives you've considered**
I can imagine only one aspect of the feature that could be an issue : if the user types a string in the **filter** `line edit` and he forgets it and his action. So he will get a shorter list of documents that could leads him to think some documents are lost.

If we think that this risk could exist, we can set the `filter text` temporary, for example:
- At each time we quit the `document mode`, we clear the `filter lineedit`.
- After X seconds, we clear the `filter lineedit`.

**Additional context**
I did not find any topic about this feature on [the internal project page](https://github.com/orgs/OpenBoard-org/projects/1).

**Comment by sebojolais** [comment]
I did not create a pull request before getting some feedback.
But the feature is available on [this branch](https://github.com/sebojolais/OpenBoard/tree/filteringDocuments).

**Comment by letsfindaway** [comment]
Nice feature! And the implementation is so easy. I was just thinking whether a regular expression as you used it in the implementation is the right thing. Most users don't know much about RegExp, so a wildcard string could also be an option. But this are already the details.

**Comment by letsfindaway** [comment]
> I can imagine only one aspect of the feature that could be an issue : if the user types a string in the **filter** `line edit` and he forgets it and his action. So he will get a shorter list of documents that could leads him to think some documents are lost.

At least we could set the property `clearButtonEnabled` on the `QLineEdit`. This is what most filters do. I would also support to clear the filter when leaving document mode, but I would not use a timer. I'm sure this would always trigger in the wrong moment.

**Comment by sebojolais** [comment]
> Nice feature! And the implementation is so easy.

Yes, easy to implement it because everything was there. I had just to connect them together.

>  I was just thinking whether a regular expression as you used it in the implementation is the right thing. Most users don't know much about RegExp, so a wildcard string could also be an option. But this are already the details.

The Qt API to filter: 
| methods | Qt 5 | Qt 6 |
| --- | --- | --- |
| setFilterWildcard | [x] | [x] |
| setFilterRegularExpression | [x] | [x] |
| setFilterFixedString | [x ] | [x] But the behavior is not consistent between versions |
| setFilterRegExp | [x] | [ ] |

In order to support Qt5 and Qt6, so we can use only `setFilterWildcard` and `setFilterRegularExpression`.
I choose `setFilterRegularExpression` because it works as `setFilterWildcard` and it can be more powerful thanks to regular expression. There is a sentence in french to say : if you can do more, you can also do lesser. Like one [KDE slogan](https://dot.kde.org/2017/05/30/plasma-510-simple-default-powerful-when-needed) : simple by default, powerful when needed.

> At least we could set the property clearButtonEnabled on the QLineEdit. This is what most filters do.

Yes, I will do the modification to provide the clear button.

> I would also support to clear the filter when leaving document mode, but I would not use a timer. I'm sure this would always trigger in the wrong moment.

I agree.

**Comment by letsfindaway** [comment]
> setFilterFixedString 	[x ] 	[x] But the behavior is not consistent

Can you please explain what inconsistent behavior this has? I could not find anything problematic in the documentation.

**Comment by sebojolais** [comment]
It was a mistake in my side because the implementation of the `filterAcceptsRow()` was too specialized for `QRegularExpression`. [Your comment](https://github.com/OpenBoard-org/OpenBoard/pull/1142#discussion_r1808145398) fixes it.
And because `setFilterFixedString()` is implemented in [Qt 6 with `QRegularExpression`](https://github.com/qt/qtbase/blob/6.7.2/src/corelib/itemmodels/qsortfilterproxymodel.cpp#L2832) and with [`QRegExp` in Qt 5](https://github.com/qt/qtbase/blob/5.15/src/corelib/itemmodels/qsortfilterproxymodel.cpp#L2875), `setFilterFixedString()` did not work with Qt5.

So this comment was outdated, now `setFilterFixedString()` works fine with Qt5 and Qt 6. Thanks to your contribution ;-)

**Comment by sebojolais** [comment]
I added a commit to implement the clearing of the filter when the `document mode` is `enable`.
